src/modules/commander/esc_calibration.cpp

```diff
@@ -75,11 +78,97 @@ bool check_battery_disconnected(orb_advert_t *mavlink_log_pub)
 	return false;
 }
 
-int do_esc_calibration(orb_advert_t *mavlink_log_pub)
+static void set_motor_actuators(uORB::Publication<actuator_test_s> &publisher, float value, bool release_control)
+{
+	actuator_test_s actuator_test{};
+	actuator_test.timestamp = hrt_absolute_time();
+	actuator_test.value = value;
+	actuator_test.action = release_control ? actuator_test_s::ACTION_RELEASE_CONTROL : actuator_test_s::ACTION_DO_CONTROL;
+	actuator_test.timeout_ms = 0;
+
+	for (int i = 0; i < actuator_test_s::MAX_NUM_MOTORS; ++i) {
+		actuator_test.function = actuator_test_s::FUNCTION_MOTOR1 + i;
+		publisher.publish(actuator_test);
+	}
+}
+
+int do_esc_calibration_ctrl_alloc(orb_advert_t *mavlink_log_pub)
+{
+	// check safety
+	uORB::SubscriptionData<safety_s> safety_sub{ORB_ID(safety)};
+	safety_sub.update();
+
+	if (safety_sub.get().safety_switch_available && !safety_sub.get().safety_off) {
+		calibration_log_critical(mavlink_log_pub, CAL_QGC_FAILED_MSG, "Disable safety first");
+		return PX4_ERROR;
+	}
+
+	int	return_code = PX4_OK;
+	uORB::Publication<actuator_test_s> actuator_test_pub{ORB_ID(actuator_test)};
+	// since we publish multiple at once, make sure the output driver subscribes before we publish
+	actuator_test_pub.advertise();
+	px4_usleep(10000);
+
+	// set motors to high
+	set_motor_actuators(actuator_test_pub, 1.f, false);
+	calibration_log_info(mavlink_log_pub, "[cal] Connect battery now");
+
+
+	uORB::SubscriptionData<battery_status_s> batt_sub{ORB_ID(battery_status)};
```

Comment:
I think we need to find a safe way to do this without the battery dependency, it's too limiting on a lot of setups.

Reply:
True

Reply:
Still relevant